qubes/events.py

```diff
@@ -115,6 +115,17 @@ def add_handler(cls, event, func):
         # pylint: disable=no-member
         cls.__handlers__[event].add(func)
 
+    @classmethod
+    def remove_handler(cls, event, func):
+        '''Remove event handler from subject's class.
+
+        :param str event: event identificator
+        :param collections.Callable handler: handler callable
+        '''
+
+        # pylint: disable=no-member
+        cls.__handlers__[event].remove(func)
+
```

Comment:
I'd write QubesDaemonEventDispatcher class, which would be instantiated once in qubes.tools.qubesd:main and passed as argument to constructor of QubesDaemonProtocol (see that partial() in loop.create_unix_server(). The mgmt.Event handler would register its connection as the outgoing event channel.

Quick round of affected methods, names are not final:

```python
class QubesDaemonEventDispatcher(object):
  def __init__(self, app):
    self.app = app
    self._connections = collections.defaultdict(list)
    type(self.app).add_handler('*', self._handler)
    qubes.vm.BaseVM.add_handler('*', self._handler)
  
  def _handler(self, subject, event, **kwargs):
    for connection in self._connections[subject]:
      connection.send_event(subject, event, **kwargs)

  def register(self, connection, subject):
    assert connection not in self._connections[subject]
    self._connections[subject].append(connection)

  def unregister(self, connection, subject):
    if subject is not None:
      self._connections[subject].remove(connection)
    else:
      for subjcoll in self._connection.values():
        try:
          subjcoll.remove(connection)
        except ValueError:
          pass

class QubesDaemonProtocol(asyncio.Protocol):
  def __init__(self, *args, app, eventdisp, debug=False, **kwargs):
    # ...
    self.eventdisp = eventdisp

  def events_register(self, subject):
    self.eventdisp.register(self, subject)

  def events_unregister(self, subject):
    self.eventdisp.unregister(self, subject)

  def connection_lost(self):
    self.events_unregister(None)

class QubesMgmt(object):
  @api('mgmt.Event', no_payload=True)
  @asyncio.coroutine
  def event(self):
    # ...
    self.events_register(self.app if self.dest.name == 'dom0' else self.dest)
    yield from wait_for_cancel